src/Processors/QueryPlan/ReadFromMemoryStorageStep.cpp

```diff
@@ -93,29 +94,52 @@ class MemorySource : public ISource
     InitializerFunc initializer_func;
 };
 
-ReadFromMemoryStorageStep::ReadFromMemoryStorageStep(Pipe pipe_) :
-    SourceStepWithFilter(DataStream{.header = pipe_.getHeader()}),
-    pipe(std::move(pipe_))
+ReadFromMemoryStorageStep::ReadFromMemoryStorageStep(QueryPlan & query_plan_,
+                                                     const SelectQueryInfo & query_info_,
+                                                     ContextPtr context_,
+                                                     const Names & columns_to_read_,
+                                                     const StorageSnapshotPtr & storage_snapshot_,
+                                                     const size_t num_streams_,
+                                                     const bool delay_read_for_global_sub_queries_) :
+    SourceStepWithFilter(DataStream{}),
+    query_plan(query_plan_),
+    query_info(query_info_),
+    context(context_),
+    columns_to_read(columns_to_read_),
+    storage_snapshot(storage_snapshot_),
+    num_streams(num_streams_),
+    delay_read_for_global_sub_queries(delay_read_for_global_sub_queries_)
 {
 }
 
 void ReadFromMemoryStorageStep::initializePipeline(QueryPipelineBuilder & pipeline, const BuildQueryPipelineSettings &)
 {
-    // use move - make sure that the call will only be made once.
+    auto pipe = makePipe();
+    if (hasOutputStream())
+    {
+        output_stream.value().header = pipe.getHeader();
+    }
+
+    if (pipe.empty())
+    {
+        auto header = storage_snapshot->getSampleBlockForColumns(columns_to_read);
+
+        // @TODO discuss about query_plan and query_info lifetimes.
+        InterpreterSelectQuery::addEmptySourceToQueryPlan(query_plan, header, query_info, context);
```

Comment:
let's remove `InterpreterSelectQuery::addEmptySourceToQueryPlan`
For this storage, we can init pipe like
```
pipe = Pipe(std::make_shared<NullSource>(header));
````

(Get a header from `output_stream->header` which should be filled in ctor)

Reply:
ok